icu4c/source/test/intltest/alphaindextst.cpp

```diff
@@ -756,4 +757,52 @@ void AlphabeticIndexTest::checkHasBuckets(const Locale &locale, UScriptCode scri
             uscript_getScript(bucket->getLabel().char32At(0), errorCode));
 }
 
+void AlphabeticIndexTest::checkMaxLabelCount() {
+    IcuTestErrorCode errorCode(*this, "checkMaxLabelCount");
+    UErrorCode status = U_ZERO_ERROR;
+    Locale locale("en");
+    UnicodeString loc = locale.getName();
+    AlphabeticIndex aindex(locale, status);
+
+    int32_t defaultMax = aindex.getMaxLabelCount();
+    assertEquals("Expect max count = 99", 99, defaultMax);
+
+    aindex.setMaxLabelCount(237, status);
+    if (U_FAILURE(errorCode)) {
+      dataerrln("%s %d  Error in index setMaxLabelCount",  __FILE__, __LINE__);
+      return;
+    }
+
+    // Verify that all bucket are cleared after setting max bucket count.
+    const UnicodeString bucketLabel = aindex.getBucketLabel();
+    assertEquals("Expect empty string", "", bucketLabel);
+
+    int32_t maxChanged = aindex.getMaxLabelCount();
+    assertEquals("Expect max count = 237", 237, maxChanged);
+
+    int32_t bigMax = 100000;
+    aindex.setMaxLabelCount(bigMax, errorCode);
```

Comment:
I don't know that it's useful to set it more than once without doing anything interesting in between.

I suggest you
- set it once to a value a bit higher than 1000
- add almost that many labels
- buildImmutableIndex() and check that you got all of the expected buckets
- add a bunch more labels, so that you exceed the max
- buildImmutableIndex() and check that the number of buckets does not exceed the max, and that the buckets you get are a  subset of those you set